### ISSUE TITLE
Suggestion: new sniff to detect code inefficiencies

### ISSUE THREAD

**Issue by jrfnl** [WordPress-Coding-Standards/WordPress-Coding-Standards#1315]
What's the general opinion on having a sniff to detect the following type of code patterns and throw `warning`s when detected ? Useful or not ?

```php
echo sprintf( 'string %s', 'replacement'); // Should use `printf()` instead.
echo esc_html__( 'string', 'domain' ); // Should use `esc_html_e()` instead.
echo selected( $selected, $current, false ); // Should use `selected( $selected, $current )` instead.
echo get_search_form( false ); // Should use `get_search_form()` instead.
```
Of course, it would need to be verified that such a call is not followed by a concatenation adding more text, but that shouldn't be too hard to do.

```php
echo sprintf( 'string %s', 'replacement') . 'something else'; // This is fine.
```

For a list of functions which have a `$echo = true` toggle as one of the parameters, [Google Search](https://www.google.nl/search?dcr=0&ei=0ZyXWuDiB8nMwQL0ha7wDA&q=developer.wordpress.org%3A+%24echo+%3D+true&oq=developer.wordpress.org%3A+%24echo+%3D+true&gs_l=psy-ab.3...58702.58702.0.59346.1.1.0.0.0.0.57.57.1.1.0....0...1.1.64.psy-ab..0.0.0....0.mu7c3a0PLtw) will probably work better than the wp.org search.

Oh... and if you like the idea, more examples of code patterns to listen for would be welcome!

**Comment by jrfnl** [comment]
@GaryJones I'd initially like to focus this ticket on PHP / WP Core constructs.

> I'd like the option to be able to extend the list of suggested replacements.

I will consider that for a later stage as that will not be easy of their are various variations of code patterns to consider and I'm not all that sure, we could allow for that when accepting input from the ruleset.

**Comment by jrfnl** [comment]
Related #346

**Comment by kkmuffme** [comment]
I think the use of (s)printf when no argument is used could be added.

e.g.
```
sprintf( __( 'hello', 'textdomain' ) );
should be:
__( 'hello', 'textdomain' );
```

also applies to printf or when esc_html__ (or esc_html_x,...) is used

**Comment by jrfnl** [comment]
For PHP functions, I believe this should be handled by a (new) sniff to be added to PHPCSExtra. For WordPress specific situations, a new sniff in WPCS is needed.
The sniff added in #1777 is a good example of that type of sniff.
